### ISSUE TITLE
LG-261: MicroProfile Metrics 2.3 Implementation

### ISSUE THREAD

**Issue by donbourne** [OpenLiberty/open-liberty#9985]
LG-261 follow from LG-262 (SPEC + API + TCK) regarding the MicroProfile Metrics 2.3 release

The MicroProfile Metrics 2.3 release includes two new major features:
1. New metric SimpleTimer (@SimplyTimed)
- Light-weight alternative to the existing Timer metric; only tracks and reports values regarding total elapsed time and invocation count

2. Metrics derived from tracking REST resource method usage (GET, POST, PUT,DELETE, OPTIONS,PATCH,HEAD)
- Backed by the SimpleTimer metric it will track elapsed time duration and invocation count
- Enabled with use of JAX-RS 2.1 (and if JAXRS methods are used). We will leverage the JAX-RS filter tracking used in APSFOUND-56

As as consequence of  the second (REST metric) feature , mpMetrics-2.3 will now depend on monitor-1.0 feature.

WAD/UFO : https://ibm.ent.box.com/file/602205571172

**Comment by atosak** [comment]
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)
 
Instructions:
 - Do the actions below and mark them complete in the checklist when they are done.
 - Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **Before Development Starts or 8 weeks before Onboarding** 
- [ ] POC Design / WAD Review Scheduled (David Chang) or N/A.
- [ ] POC Design / WAD Reviewed (Feature Owner) or N/A.
- [ ] Complete any follow-ons from the POC Review.
- [ ] Design / WAD Approval (Alasdair Nottingham) or N/A.
- [ ] No Design / No WAD Approval (Arthur De Magalhaes - cloud / Alasdair Nottingham - server) or N/A.
- [ ] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [ ] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [ ] Create a child task of the epic entitled "FAT Approval Test Summary".  Add and fill in the template as described here: https://github.ibm.com/was-liberty/WS-CD-Open/wiki/Feature-Review-(Feature-Test-Summary-Process)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **3 weeks before Onboarding** 
- [ ] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE ** 3 weeks before Onboarding** 
- [ ] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **2 weeks before Onboarding** 
- [ ] Implementation complete. (Epic owner / Feature owner) 
- [ ] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## APPROVALS with TARGET COMPLETION DATE **2 to 1 week before Onboarding** 
Prereq: You must have the Design Approved or No Design Approved label on the GitHub Epic.
- [ ] **Accessibility** - (G Scott Johnston). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [ ] **FAT Liberty SOE** - (Kevin Smith). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [ ] **Globalization** (Sam Wong - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.
- [ ] **ID** - (Kareen Deen). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [ ] **Performance** - (Jared Anderson). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [ ] **Serviceability** - (Don Bourne). Serviceability has been addressed. 
- [ ] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [ ] **SVT** - (Greg Ecock - Cloud, Brian Hanczaryk- APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.
- [ ] **Demo** - (Liberty only - Tom Evans or Chuck Bridgham). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## TARGET COMPLETION DATE **1 week before Onboarding** 
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## NOT REQUIRED FOR A FEATURE
- [ ] **OL Guides** - (Yee-Kang Chang). Assessment for OL Guides is complete or N/A. 
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
 
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Related Deliverables TARGET COMPLETION DATE **General Availability** 
- [ ] Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by Channyboy** [comment]
WAD/UFO : https://ibm.ent.box.com/file/602205571172

**Comment by Channyboy** [comment]
https://ibm.box.com/s/4vmex4gxj28y74s4grnn7p0hczrr4105

**Comment by NottyCode** [comment]
Feedback on design:

1. Please use the new design template for future designs
2. Slide 26 uses rest.request but earlier and later in the deck it uses REST.request, is this correct or is it a typo
3. Slide 28 references installUtility which doesn't exist in Open Liberty so should not be there.
4. Slide 29, the first bullet is likely unnecessary.
5. Slide 35, I would prefer to see some statement about the performance effect rather than "it will have one" are we talking 5%, 10% more?  Is anything being done to address this? It should be clear that this is vs the mpMetrics-3.2 feature and isn't a degradation in performance for an existing feature.
6. Slide 36 doesn't talk about servicability, it talks about mitigating an unqualified performance impact, so really should be on slide 35.

**Comment by Channyboy** [comment]
@NottyCode 
Slide 26 is a typo. It should be `REST.request` and is now fixed.

Slide 28 and 29 fixed. 

Slide 35 updated with statement.

Slide 36 modified to address customer concern regarding additional PMI stats (similar to APSFOUND-56)

**Comment by Channyboy** [comment]
Slide 35 updated with performance results from performance team.

**Comment by Channyboy** [comment]
@gscottj There are no addition of any new or change to any existing UI. May I get your approval on Accessibility? Thanks!

**Comment by gscottj** [comment]
No new user interfaces are being added nor any changes to existing user interfaces. No required accessibility verification testing.

**Comment by Channyboy** [comment]
Further updates to WAD.

slide 35 - Clarified further performance impact.

slide 36 - Activation of PMI stats are classified more-so as performance and has been removed.

**Comment by Channyboy** [comment]
@samwatibm Hi Sam, translation files have been merged. May you approve.

**Comment by Channyboy** [comment]
@chirp1 
Doc issues are opened with this item:
https://github.com/OpenLiberty/docs/issues/1109

May you approve for ID.

**Comment by Channyboy** [comment]
@sabolo May you review?
Feature  Test Summary: https://github.com/OpenLiberty/open-liberty/issues/10946

**Comment by Channyboy** [comment]
@donbourne 
Serviceability Approval Comment - Please answer the following questions for serviceability approval:

UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?

N/A - no problems expected to be encountered with new metric and metric data. Underlying logic behaves the same as other existing metrics (i.e registration, retrieval and export/display).

2. Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).
a) What problem paths were tested and demonstrated?
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

N/A - no problems expected to be encountered with new metric and metric data. Underlying logic behaves the same as other existing metrics (i.e registration, retrieval and export/display).


3. SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
Andrew Reider and Tanya Kulik

b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
Yes.
I agree with 3b for mpMetrics-2.3 , on the basis that the new feature is similar in complexity to the previous version with regard to problem resolution. Both versions are publicly documented (and most of it is the same), so developers/clients using either can potentially resolve issues on their own or using solutions that worked for version 2.2

Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.
WAS L3: Liberty Log Analytics

**Comment by Channyboy** [comment]
@tevans78 Demo is scheduled for Liberty 20.5 - March 10th EOI.

**Comment by skasund** [comment]
L2 has requested STE slides for this feature.  The STE template can be found at the link below.  You can use either one to create the education.  

Slide Template:  https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Github Template:  https://pages.github.ibm.com/WASL3/site/STE/about

Please upload the completed slides to the same STE Archive BOX folder.  Thanks!

**Comment by Channyboy** [comment]
@skasund STE slide here : https://ibm.ent.box.com/file/629590802249

**Comment by skasund** [comment]
@Channyboy   Thanks for the slides.  I've approved the feature.

**Comment by chirp1** [comment]
The writers are working down a prioritized backlog. As I mentioned to David Chan, the doc updates for OL likely won't be made by the time 20.0.0.4 goes out. David will have a GA log post for the Epic. Approving.

**Comment by Channyboy** [comment]
mpMetrics-2.3 moved to GA
See https://github.com/OpenLiberty/open-liberty/pulls?q=is%3Apr+tevan+is%3Aclosed
closing.
